### PR TITLE
Fix metrics stats for VMs not running

### PR DESCRIPTION
**Pull request by joseflauzino**
### Description

This PR fixes the behavior of the stats collector with respect to VMs that are not running. The need for this is that if a running VM has a state change for some reason, ACS keeps showing the latest metrics stats collected.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

Using ACS UI, I performed the following steps:
- I deployed a VM and waited for the complete boot;
- I enabled the presentation of metric statistics in the UI;
- I waited for ACS to collect and show metrics stats;
- And then, I tested two different cases:
	- Case 1 - I stopped VM using ACS UI;
	- Case 2 - I simulated a VM crash by manually stopping it (directly from the hypervisor);
- In both cases, the metrics stats for the VM were successfully cleared. As expected, when I tested Case 1, cleanup was done immediately. In Case 2, I had to wait for ACS to identify that the VM was stopped and the stats collector try to collect new metric stats.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 1]
```diff
@@ -1485,6 +1487,32 @@ private void storeVirtualMachineStatsInMemory(VmStatsEntry statsForCurrentIterat
...
+        s_logger.debug(String.format("Removing stats from VM with ID: %s .",vmId));
+        _VmStats.remove(vmId);
+    }
+
+    /**
+     * Removes stats of virtual machines that are not running from memory.
+     */
+    protected void cleanUpVirtualMachineStats() {
+        List<Long> allRunningVmIds = new ArrayList<Long>();
+        for (UserVmVO vm : _userVmDao.listAllRunning()) {
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/server/StatsCollector.java:1504]: VM stats are captured for running VMs only
https://github.com/apache/cloudstack/blob/a78af608144f80649c409ebd933265d2d755f289/server/src/main/java/com/cloud/server/StatsCollector.java#L579

 _removeVirtualMachineStats()_  removes the vm stat if the VM is stopped/destroyed after adding to the stats, so no need for this cleanup method.

Comment (joseflauzino) [reply server/src/main/java/com/cloud/server/StatsCollector.java:1504]: Thanks for the review @sureshanaparti!
The `cleanUpVirtualMachineStats()` method is important to maintain consistency of the metric stats shown by ACS even when there are cases of VMs changing state unexpectedly (e.g., when a VM crashes; or when the VM is stopped directly by the hypervisor interface, etc). In these cases the `removeVirtualMachineStats()` method is not executed, so the `cleanUpVirtualMachineStats()` method does its job.

**Comment by joseflauzino** [comment]
Any update about this PR?

**Comment by DaanHoogland** [comment]
code looks good, let's retest
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by joseflauzino** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

**Review by GutoVeronezi — APPROVED** [review]
LGTM

@DaanHoogland can we merge this?

**Comment by DaanHoogland** [comment]
@GutoVeronezi yes, I think @sureshanaparti's question was answered to satisfaction as well. let's merge.

**Comment by sureshanaparti** [comment]
@joseflauzino Is this good to go in 4.16.1? If so, please rebase with '4.16'.

**Comment by joseflauzino** [comment]
@sureshanaparti for me we can keep it in the main (4.17).

**Comment by GutoVeronezi** [comment]
Merging this based on 2 LGTMs and [package](https://github.com/apache/cloudstack/pull/5633#issuecomment-974081799) and [test](https://github.com/apache/cloudstack/pull/5633#issuecomment-975810322)  results.

**Pull request merged**